File: docs/ASL/MapGenerators/Hex Generators/src/lib/hex-geom.js
// lib/hex-geom.js
// Flat-top hex geometry utilities (N = 0 side indexing).
// Compatible with the visualizer, viewer, and generator.
//
// Coordinate system: SVG-style (x right, y down).
// Hex orientation: flat-top. Side names are clockwise: ["N","NE","SE","S","SW","NW"].

//////////////////////////
// Constants & Indexing //
//////////////////////////

/** Default side order, clockwise, with N=0. */
export const SIDE_ORDER = ["N", "NE", "SE", "S", "SW", "NW"];

/**
 * Outward-normal angles (radians) at each hex side midpoint (local hex coords).
 * Values are chosen for flat-top orientation with SVG's Y+ downward.
 */
export const EDGE_ANGLE = {
  N: 3 * Math.PI / 2,   // up
  NE: 11 * Math.PI / 6, // up-right
  SE: Math.PI / 6,      // down-right
  S: Math.PI / 2,       // down
  SW: 5 * Math.PI / 6,  // down-left
  NW: 7 * Math.PI / 6,  // up-left
};

/**
 * Get index of a side name inside a given order (default SIDE_ORDER).
 * @param {string} name - One of "N","NE","SE","S","SW","NW".
 * @param {string[]} [order=SIDE_ORDER]
 * @returns {number} index or -1 if not present
 */
export function idxOf(name, order = SIDE_ORDER) {
  return order.indexOf(name);
}

/**
 * Normalize an index to [0..5] and return the side name from the given order.
 * @param {number} index - Can be negative or overflow.
 * @param {string[]} [order=SIDE_ORDER]
 * @returns {string} side name
 */
export function nameOf(index, order = SIDE_ORDER) {
  return order[((index % 6) + 6) % 6];
}

/**
 * Opposite side index (3 steps away).
 * @param {number} i
 * @returns {number}
 */
export function oppositeIndex(i) {
  return (i + 3) % 6;
}

//////////////////////////
// Vertex/Edge Geometry //
//////////////////////////

/**
 * Get vertex position (local) at unit radius, scaled by `size`.
 * Vertex 0 is at angle 0 (to the right), then 60° steps (flat-top).
 * @param {number} i - vertex index [0..5]
 * @param {number} size - hex radius
 * @returns {{x:number,y:number}}
 */
export function getVertexPosition(i, size) {
  const a = (Math.PI / 3) * i; // 0,60,120,...
  return { x: size * Math.cos(a), y: size * Math.sin(a) };
}

/**
 * Get the midpoint of hex side `edgeIndex` (local), at radius `size`.
 * Computed as midpoint between vertices i and (i+1).
 * @param {number} edgeIndex - [0..5]
 * @param {number} size - hex radius
 * @returns {{x:number,y:number}}
 */
export function getEdgeMidpoint(edgeIndex, size) {
  const v1 = getVertexPosition(edgeIndex, size);
  const v2 = getVertexPosition((edgeIndex + 1) % 6, size);
  return { x: (v1.x + v2.x) / 2, y: (v1.y + v2.y) / 2 };
}

/**
 * Apothem (center to side distance) for a flat-top hex of radius `size`.
 * @param {number} size
 * @returns {number}
 */
export function apothem(size) {
  return size * Math.cos(Math.PI / 6);
}

/**
 * Linear interpolation between two points.
 * @param {{x:number,y:number}} p1
 * @param {{x:number,y:number}} p2
 * @param {number} t - [0..1]
 * @returns {{x:number,y:number}}
 */
export function lerpPoint(p1, p2, t) {
  return { x: p1.x + (p2.x - p1.x) * t, y: p1.y + (p2.y - p1.y) * t };
}

/////////////////////
// Drawing Helpers //
/////////////////////

/**
 * Points string for a hex polygon centered at (cx,cy) with radius `size`.
 * Suitable for the SVG <polygon points="..."> attribute.
 * @param {number} cx
 * @param {number} cy
 * @param {number} size
 * @returns {string}
 */
export function hexPolygonPoints(cx, cy, size) {
  const pts = [];
  for (let i = 0; i < 6; i++) {
    const v = getVertexPosition(i, size);
    pts.push(`${(cx + v.x).toFixed(3)},${(cy + v.y).toFixed(3)}`);
  }
  return pts.join(" ");
}

/**
 * Outward normal angle (degrees) for side i (local coords).
 * Useful for label/marker orientation if needed.
 * @param {number} i - side index [0..5]
 * @param {number} size - hex radius (unused here; included for parity)
 * @returns {number} angle in degrees
 */
export function edgeNormalAngleDeg(i, size) { // size kept for API symmetry
  const m = getEdgeMidpoint(i, size || 1);
  return (Math.atan2(m.y, m.x) * 180) / Math.PI;
}

File: src/Legacy/Samples/ASL/AslHexMap.Core/Geometry/HexGeom.cs
using System;
using System.Globalization;
using System.Text;
using AslHexMap.Core.Schema; // for Side

namespace AslHexMap.Core.Geometry
{
    public static class HexGeom
    {
        /// <summary>Flat-top hex vertices, clockwise, starting at angle 0° (East).</summary>
        public static (double x, double y)[] PointsFlatTop(double cx, double cy, double size)
        {
            var pts = new (double x, double y)[6];
            for (int i = 0; i < 6; i++)
            {
                var angleDeg = 60 * i; // 0,60,120,180,240,300
                var angleRad = Math.PI / 180 * angleDeg;
                pts[i] = (cx + size * Math.Cos(angleRad), cy + size * Math.Sin(angleRad));
            }
            return pts;
        }

        public static string ToSvgPoints((double x, double y)[] pts)
        {
            var sb = new StringBuilder();
            var inv = CultureInfo.InvariantCulture;
            for (int i = 0; i < pts.Length; i++)
            {
                if (i > 0) sb.Append(' ');
                sb.Append(pts[i].x.ToString("0.###", inv));
                sb.Append(',');
                sb.Append(pts[i].y.ToString("0.###", inv));
            }
            return sb.ToString();
        }

        /// <summary>
        /// Midpoint of a hex side for flat-top orientation (0=N,1=NE,2=SE,3=S,4=SW,5=NW).
        /// </summary>
        public static (double x, double y) EdgeMidpointFlatTop(double cx, double cy, double size, Side side)
        {
            var p = PointsFlatTop(cx, cy, size);
            // Map side -> (v1,v2) index in our points array [E(0), NE(1), NW(2), W(3), SW(4), SE(5)]
            (int a, int b) = side switch
            {
                Side.N => (1, 2), // NE-NW
                Side.NE => (0, 1), // E-NE
                Side.SE => (5, 0), // SE-E
                Side.S => (4, 5), // SW-SE
                Side.SW => (3, 4), // W-SW
                Side.NW => (2, 3), // NW-W
                _ => (1, 2)
            };
            return ((p[a].x + p[b].x) * 0.5, (p[a].y + p[b].y) * 0.5);
        }

        /// <summary>Label anchor near the NW interior of the hex.</summary>
        public static (double x, double y) LabelAnchorNW(double cx, double cy, double size)
        {
            var p = PointsFlatTop(cx, cy, size);
            var nw = p[2]; // NW vertex
            // Move inwards ~25% toward center so the label sits inside the hex
            var x = cx + (nw.x - cx) * 0.75;
            var y = cy + (nw.y - cy) * 0.75;
            return (x, y);
        }
    }
}
